Divesoft: Fix the Log File Import and Add New Liberty Model.

mikeller opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [x] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
Fix the import of Divesoft log files to default to the Freedom model if
an unknown model identifier is encountered.
Also add support for the 'Liberty 2021' model identifier.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
1. fix the import of Divesoft log files to default to the Freedom model if an unknown model identifier is encountered;
2. add support for the 'Liberty 2021' model identifier.

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->
Fixes #4536.

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->

glance- commented:
I'd still prefer #4510 which solves this issue in a more future proof and less fragile way.

mikeller commented:
> I'd still prefer #4510 which solves this issue in a more future proof and less fragile way.

If you are able to supply explanations for what you consider to be 'not future proof' and 'fragile' in the proposed approach I am happy to consider amendments to address them.

glance- commented:
This still assumes that the DLF-files are in V2 format, without verifying that. It would be better to just leave the whole file parsing to libdivecomputer, which already supports both V1 and V2 files.

Also, I think it's unnecessary to emit a error if the bytes checked doesn't match  one of the known prefixes. Just leave that parsing to libdivecomputer, which already handles it.

mikeller commented:
@glance-:

> This type of threats is not productive.

This is not a treat - I am just highlighting a potential consequence of your behaviour.

> If you don't understand my arguments, fine, but making assumptions and threats are really harmful for this community.

The problem is that you are not making reasoned arguments, you are just making unfounded allegations and then not backing them when asked to.
And yes, I think that your toxic behaviour is harmful.

glance- commented:
There are two cases reported, https://groups.google.com/g/subsurface-divelog/c/dsr0mPdfH3U/m/kDPpnNl3AAAJ and https://groups.google.com/g/subsurface-divelog/c/WOP00QlYOWM/m/4Nc6XatlDgAJ where this code that assumes all files are DLF v2 and trying to match serial numbers have failed for users. This clearly shows the fragility of this approach.

The other argument is that this code is unnecessary, because the actual parsing is done by libdivecomputer. Just removing these assumptions based on serial number will solve these and and future changes and added ranges of serial numbers.

If we would like to do any additional parsing of the DLF content to augment the libdivecomputer parsing, either add it to libdivecomputer or add it after the libdivecomputer parsing of the dive file, and don't reject the file just because the data doesn't match subsurface assumptions of how the data looks.

I'd would have just excluded any model name from the model string, and just write something like:
> d->dcs[0].model = devdata.vendor + " (Imported from file)";

mikeller commented:
> There are two cases reported, [groups.google.com/g/subsurface-divelog/c/dsr0mPdfH3U/m/kDPpnNl3AAAJ](https://groups.google.com/g/subsurface-divelog/c/dsr0mPdfH3U/m/kDPpnNl3AAAJ) and [groups.google.com/g/subsurface-divelog/c/WOP00QlYOWM/m/4Nc6XatlDgAJ](https://groups.google.com/g/subsurface-divelog/c/WOP00QlYOWM/m/4Nc6XatlDgAJ) where this code that assumes all files are DLF v2 and trying to match serial numbers have failed for users. This clearly shows the fragility of this approach.

You are incorrect. None of the cases that you reference applies to the code in this pull request. Please stop spreading lies.

> The other argument is that this code is unnecessary, because the actual parsing is done by libdivecomputer. Just removing these assumptions based on serial number will solve these and and future changes and added ranges of serial numbers.

You are incorrect again - currently the parser in libdivecomputer does not try to detect the correct type of Divesoft dive computer: https://github.com/subsurface/subsurface/pull/4510#issuecomment-2824491678

mikeller commented:
> The list of serial number prefixes in your proposed fix is still incomplete. There are at least two Freedom models missing. While we could easily add those, the list will become out of date again in the future when Divesoft adds some more new models.

Can you please point me to the user reports about these, I am keen to add them.

> That's the future proof problem @glance- is referring to: your code will need an update whenever a new serial number becomes available, while the parsing itself doesn't need any changes.

I really can not follow your argument here - the code proposed in #4510 is just as much 'not future proof' - when a new model ID becomes available it will NOT give the user the correct model name - just like the code in this pull request. The only difference is that this code is giving the user the information in the cases where we have it, instead of just ignoring it.

> Your proposed fix is an improvement in the sense that it will no longer fail to import V1 files, but the model detection will still not always be accurate. It's still completely wrong for V1 files (and will even crash the application when trying to import a V1 file that only contains the header) and also doesn't work for V2 files without a serial number.

You seem to be very worried about supporting V1 files - I am happy to investigate support for properly detecting them. Can you please point me to where I can obtain the software that is generating them, or how a current diver would obtain them?
And the same goes for files without a serial number - how can a diver generate one of these?

> @glance- and I consider it a major bug when a DLF file, which can be parsed by libdivecomputer, fails to import in subsurface. Being able to display the correct dive computer model is less critical for us and is at most a small cosmetic bug.

I think this is a difference in perspective - as a technical diver most of my dives will have 2 or even 3 profiles from different dive computers. So being able to easily distinguish which dive computer's profile I am looking at is way more than a 'cosmetic issue' for me as a technical diver.
And I suspect that, if anything, this will be exacerbated by the fact that in all likelihood, a lot of divers using the Divesoft Liberty CCR will be using Divesoft Freedom+ dive computers, in order to take advantage of the unified UI between CCR and dive computer.

jefdriesen commented:
> Can you please point me to the user reports about these, I am keen to add them.

There are no user reports. It's based on info I received from Divesoft. The missing ones are 7011 and 7044.

> I really can not follow your argument here - the code proposed in #4510 is just as much 'not future proof' - when a new model ID becomes available it will NOT give the user the correct model name - just like the code in this pull request. The only difference is that this code is giving the user the information in the cases where we have it, instead of just ignoring it.

In the current code, the model number is used for two different purposes: for passing to the parser and for displaying the model name. Since it's not needed for the first case, I proposed to remove that code. Code that isn't there, doesn't need updates. @glance- made an extra suggestion to take care of the display name.

I don't mind trying to do the detection anyway, but at the moment that code doesn't handle the V1 format.

> You seem to be very worried about supporting V1 files - I am happy to investigate support for properly detecting them.

I think it's important to keep supporting V1 files. Remember not so long ago, the V1 format was the only format subsurface supported. The V2 format is still relative new and there will still be plenty of users with V1 files. See the bug reports that were already mentioned earlier.

> Can you please point me to where I can obtain the software that is generating them, or how a current diver would obtain them? And the same goes for files without a serial number - how can a diver generate one of these?

They are generated by all Divesoft dive computers which haven't upgraded to the new format. I don't even know if all models can upgrade to newer firmware with the V2 format, so it's possible that some users with an older model are stuck with V1 forever.

The missing serial number should only happen for the factory test dives, so that's less critical.

> > @glance- and I consider it a major bug when a DLF file, which can be parsed by libdivecomputer, fails to import in subsurface. Being able to display the correct dive computer model is less critical for us and is at most a small cosmetic bug.
> 
> I think this is a difference in perspective - as a technical diver most of my dives will have 2 or even 3 profiles from different dive computers. So being able to easily distinguish which dive computer's profile I am looking at is way more than a 'cosmetic issue' for me as a technical diver. And I suspect that, if anything, this will be exacerbated by the fact that in all likelihood, a lot of divers using the Divesoft Liberty CCR will be using Divesoft Freedom+ dive computers, in order to take advantage of the unified UI between CCR and dive computer.

Sure, I certainly understand your argument. But that doesn't make the display name more important than being able to import the dive in the first place! Right now users simply can't import their V1 files at all. I'm pretty sure that's more important to them than a correct display name.

mikeller commented:
> > Can you please point me to the user reports about these, I am keen to add them.
> 
> There are no user reports. It's based on info I received from Divesoft. The missing ones are 7011 and 7044.

Ok, I'll contact them to get a complete list.

> > You seem to be very worried about supporting V1 files - I am happy to investigate support for properly detecting them.
> 
> I think it's important to keep supporting V1 files. Remember not so long ago, the V1 format was the only format subsurface supported. The V2 format is still relative new and there will still be plenty of users with V1 files. See the bug reports that were already mentioned earlier.

Do you have an approximate idea for when Divesoft switched from V1 to V2?
I think I tested this code against all user bug reports mentioned in this pull request, and I could not find a single one that mentioned a V1 log file, or that contained a log file that could not be processed by this change - in fact all but one of them could be processed by the changes made in https://github.com/subsurface/subsurface/pull/4367.

Can you point out the specific user bug report that references a V1 file, so that I can contact the user?

> > Can you please point me to where I can obtain the software that is generating them, or how a current diver would obtain them? And the same goes for files without a serial number - how can a diver generate one of these?
> 
> They are generated by all Divesoft dive computers which haven't upgraded to the new format. I don't even know if all models can upgrade to newer firmware with the V2 format, so it's possible that some users with an older model are stuck with V1 forever.

So the V1 files can be obtained by attaching the Divesoft dive computer as a storage device and downloading the logs directly from the dive computer?

> Sure, I certainly understand your argument. But that doesn't make the display name more important than being able to import the dive in the first place! Right now users simply can't import their V1 files at all. I'm pretty sure that's more important to them than a correct display name.

Agreed. If you can point me to a user who has this problem with their current hardware I am keen to work with them to get them supported.

glance- commented:
This has been stalled for quite some time, and #4510 is ready to be merged, so my suggestion is to pick than one instead of having this stalled for any longer.

We can always revisit other alternatives in new PR's.

jakubhecko commented:
After a brief testing of the latest build I am quite happy with the results. The Subsurface parsed everything I was able to throw at it (Import / Import log files), including logs with invalid or missing serial numbers. Thanks for the fix !
> **Artifacts:** [Subsurface-Windows-6.0.5425-patch.1.pull-request.fix-divesoft-file-import-2](https://github.com/subsurface/subsurface/suites/43736391554/artifacts/3792890294) _**WARNING:** Use at your own risk._

mikeller commented:
@jakubhecko Thanks for checking. I have been able to get back to working on this, and I am preparing a change that addresses the concerns that have been raised about properly recognising V1 log files and files generated by factory test dives. This should be ready in the next week or so.

mikeller commented:
@jakubhecko: This is ready for review now - I have added some of the sample logs provided by you for testing.
